src/LanguageServer.php

```diff
@@ -141,6 +141,10 @@ public function __construct(ProtocolReader $reader, ProtocolWriter $writer)
                         $e
                     );
                 }
+
+                // When a request is processed, clear the caches of definition resolver as not to leak memory.
+                $this->definitionResolver->clearCache();
```

Comment:
I don't understand why the cache needs to be cleared after every single protocol message. This is a very hot code path that is even hit in the middle of indexing. Could you explain?

Reply:
My intent was to
* clear the cache when some user action causes the cache to be filled to avoid memory leaks, and 
* not have to worry about invalidating the cache on user edits.

You're right, the code does not fulfill that intent. I also did not realize that it could be hit in the middle of indexing - I imagine that this might cause some weird behavior if user edit is performed on a file that is being parsed.

Reply:
It shouldn't happen while a file is parsing, but whenever control is yielded back to the event loop (definitely between indexing files) or getting file contents.

The problem I have is that a cache is not useful if it is constantly cleared when it doesn't have to. For example, moving the mouse over a document can trigger thousands of hover requests within milliseconds. All of these would cause the whole cache to be cleared every time.

Reply:
Yeah - it currently only speeds things up when the same docblock is parsed multiple times on the same request. If it's moved into `PhpDocument`, this issue could also be resolved.

Reply:
@felixfbecker - I started to move this to `PhpDocument` and found that it was rather difficult to pass `PhpDocument` to all functions requiring access to docblocks.

Should I
- Pass `PhpDocument` as argument individually to all `DefinitionResolver` functions requiring it, even if it requires adding the same argument to many places.
- Have a per-document `DefinitionResolver`, containing a reference to `PhpDocument`
- Have `PhpDocumentFactory` in `DefinitionResolver`, access the document through the URI of a node. This would require some changes though, since `PhpDocument` parses in constructor, so it would not be accessible via `PhpDocumentFactory` during the initial parse.
- Something else?

Reply:
I see. Maybe create a dedicated class for getting docblocks that caches them? The only constraint is that the memory needs to be disposed when a PhpDocument is disposed.